examples/tutorial.dx

```diff
@@ -0,0 +1,483 @@
+  to arrays with _any_ index type. This is possible by declaring an `n => Int32`
+  array argument type: this declares the type variable `n` as the index type of
+  the array argument.
+
+def arrAdd5 (x : n => Int32) : n => Int32 = for i. x.i + 5
+
+:t arrAdd5 x2
+
+' But function types can help you out even more. For instance, since index types
+  are statically known, type checking can ensure that array arguments have valid
+  dimensions. This is "shape safety".
+
+' For instance, let's write a function adding two 2D arrays with the same shape:
+
+:t x
+:t y
+
+def arrAdd (x : m => n => Int32) (y : m => n => Int32) : m => n => Int32 =
+    for i. for j. x.i.j + y.i.j
+
+:t arrAdd x y
+
+:t arrAdd x (trans y)
+
+' The type system checked for us that input arrays indeed have the same shape.
+
+'## Writing loops
+
+' Dex is a functional language - but when writing mathematical algorithms,
+  it is often convenient to temporarily put aside immutability and write
+  imperative code using mutation.
+
+' For example, let's say we want to actually implement the `sum` function
+  ourselves by accumulating summed values in-place. In Python, implementing this
+  is not directly possible solely via list comprehensions, so we would write a
+  loop.
+
+'  `acc = 0`
+
+'  `for i in range(10):`
+
+'      `acc = acc + x[i]`
+
+'   `return acc`
+
+' In Dex, values are immutable, so we cannot directly perform mutation. But Dex
+  includes algebraic effects, which are a purely-functional way to modeling
+  side-effects like mutation. We can write code that looks like mutation using
+  the `State` effect, which provides getter and setter functionality (via `get`
+  and `:=` assignment). Here's what it looks like:
+
+def arrSum (x : n => Int32) : Int32 =
+    -- acc = 0
+    init = 0
+
+    -- (ignore for now)
+    withState init $ \acc.
+         -- for i in range
+         for i.
+             -- acc = acc + x[i]
+             acc := (get acc) + x.i
+         -- return acc
+         get acc
+             
+:p arrSum x2
+
+' So, even though Dex is a functional language, it is possible to write loops
+  that look similar to ones that truly perform mutation. However, there is one
+  line which is quite new and a bit scary. Let us look into that line in a bit
+  more detail.
+
+' First: `$`. This symbol is used in Dex just like it is used in Haskell, but
+  if you haven't seen it before, it seems a bit strange. `$` is the function
+  application operator: it basically replaces of expression-grouping parentheses
+  `(f x)` when it is inconvenient to write them. For example, the following two
+  expressions are identical:
+
+:t arrSum (x2 + x2)
+
+:t arrSum $ x2 + x2
+
+' Next: `\`. This symbol is the lambda sigil in Dex. It is analogous to the
+  `lambda` keyword in Python, and starts the definition of a function value
+  (i.e. closure). In `arrSum` above: the lambda takes an argument named `acc`
```

Comment:
Technically speaking it may or may not be a closure.
Idk if that kind of pedantry is useful in the tutorial though.

It might be nice to mention in this section that this can also be used to define functions, like the `def` keword above.

It also might be nice to mention the symmetry  between `for i.` and `\i.`;
though I thing that might be awkard so maybe not worth interupting flow that much.

Reply:
> Technically speaking it may or may not be a closure.
> Idk if that kind of pedantry is useful in the tutorial though.

Good note. From my Swift background, I use "closure" as a generic name for "function values", including both function references (trivial closures, capturing no context) and closure expressions (`\x. blah x`).

> It might be nice to mention in this section that this can also be used to define functions, like the `def` keword above.

I think omitting this (explanation of first-class functions, lambdas can be assigned to variables) is reasonable, given that Python programmers are familiar with this.

> It also might be nice to mention the symmetry between `for i.` and `\i.`;
> though I thing that might be awkard so maybe not worth interupting flow that much.

I agree! We discussed a bit more about "arrays as functions" in [a different thread](https://github.com/google-research/dex-lang/pull/443#discussion_r554312595). Sasha pointed out that "`for` as a list comprehension" is more user-friendly than "arrays as function types/lambdas" (which is more programming-language-theoretic) - and this tutorial seems catered towards folks coming from Python.

Reply:
I'm fine with closure. This is not supposed to be a super formal introduction, and in most informal settings lambda ≈ closure.